### PR TITLE
Enable AAD auth for EventGridPublisherClient

### LINKED ISSUE #22109: Add AAD support
Update client library to support authenticating with AAD.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/22109, https://github.com/Azure/azure-sdk-for-net/issues/22505, and https://github.com/Azure/azure-sdk-for-net/issues/22515

### CODE REVIEW

File: sdk/eventgrid/test-resources.json [hunk 1]
```diff
@@ -1,109 +1,138 @@
...
+      }
+    }
+  },
+  "variables": {
+    "apiVersion": "2020-04-01-preview",
+    "eventGridTopicName": "[concat(parameters('baseName'), 'topic')]",
+    "eventGridDomainName": "[concat(parameters('baseName'), 'domain')]",
+    "cloudeventTopicName": "[concat(parameters('baseName'), 'cloudevent-topic')]",
+    "customeventTopicName": "[concat(parameters('baseName'), 'customevent-topic')]",
+    "eventGridDataSenderRoleId": "d5a91429-5739-47e2-a06b-3470a27159e7"
```
Comment (JoshLove-msft) [inline sdk/eventgrid/test-resources.json:33]: Formatting was messed up before so I will comment on the actual changed lines.

File: sdk/eventgrid/test-resources.json [hunk 2]
```diff
@@ -1,109 +1,138 @@
...
-            "value": "[reference(variables('customeventTopicName')).endpoint]"
-        },
-        "EVENT_GRID_CUSTOM_EVENT_TOPIC_KEY": {
-            "type": "string",
-            "value": "[listKeys(resourceId('Microsoft.EventGrid/topics', variables('customeventTopicName')), variables('apiVersion')).key1]"
-        }
+      }
+    },
+    {
+      "type": "Microsoft.Authorization/roleAssignments",
```
Comment (JoshLove-msft) [inline sdk/eventgrid/test-resources.json:94]: This adds the role assignment to the service principal.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs [hunk 3]
```diff
@@ -50,6 +50,22 @@ public EventGridPublisherClient(Uri endpoint, AzureKeyCredential credential, Eve
...
+        /// <param name="options">The set of options to use for configuring the client.</param>
+        public EventGridPublisherClient(Uri endpoint, TokenCredential credential, EventGridPublisherClientOptions options = default)
+        {
+            Argument.AssertNotNull(endpoint, nameof(endpoint));
+            Argument.AssertNotNull(credential, nameof(credential));
+            options ??= new EventGridPublisherClientOptions();
+            _uriBuilder = new RequestUriBuilder();
+            _uriBuilder.Reset(endpoint);
+            _uriBuilder.AppendQuery("api-version", options.Version.GetVersionString(), true);
+            _pipeline = HttpPipelineBuilder.Build(options, new BearerTokenAuthenticationPolicy(credential, "https://eventgrid.azure.net/.default"));
```
Comment (jsquire) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:65]: Is this auth scope valid for all clouds?

**Review by jsquire — APPROVED** [review]

**Review by ellismg — APPROVED** [review]
LGTM

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:65]: Hmm, @ellismg do you happen to know the answer to this?

Comment (ellismg) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:65]: I'm suspect this is the one scope, but I'll ask the service team.

Comment (ellismg) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:65]: Spoke with the service team - the same scope is used across clouds.

**Pull request merged**